go/arrow/array/boolean_test.go

```diff
@@ -0,0 +1,258 @@
+	defer b.Release()
+
+	b.AppendValues(values, valids)
+
+	arr := b.NewArray().(*array.Boolean)
+	defer arr.Release()
+
+	if got, want := arr.Len(), len(valids); got != want {
+		t.Fatalf("got=%d, want=%d", got, want)
+	}
+
+	if got, want := arr.NullN(), 2; got != want {
+		t.Fatalf("got=%d, want=%d", got, want)
+	}
+
+	vs := make([]bool, arr.Len())
+
+	for i := range vs {
+		vs[i] = arr.Value(i)
+	}
+
+	if got, want := vs, values; !reflect.DeepEqual(got, want) {
+		t.Fatalf("got=%v, want=%v", got, want)
+	}
+
+	tests := []struct {
+		interval [2]int64
+		nulls    int
+		want     []bool
+	}{
+		{
+			interval: [2]int64{2, 9},
+			nulls:    1,
+			want:     []bool{true, false, false, false, true, false, true},
+		},
+		{
+			interval: [2]int64{0, 7},
+			nulls:    1,
+			want:     []bool{true, false, true, false, false, false, true},
+		},
+		{
+			interval: [2]int64{1, 8},
+			nulls:    2,
+			want:     []bool{false, true, false, false, false, true, false},
+		},
+		{
+			interval: [2]int64{2, 7},
+			nulls:    0,
+			want:     []bool{true, false, false, false, true},
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run("", func(t *testing.T) {
+
+			slice := array.NewSlice(arr, tc.interval[0], tc.interval[1]).(*array.Boolean)
+			defer slice.Release()
+
+			if got, want := slice.NullN(), tc.nulls; got != want {
+				t.Errorf("got=%d, want=%d", got, want)
+			}
+
+			if got, want := slice.Len(), len(tc.want); got != want {
+				t.Fatalf("got=%d, want=%d", got, want)
+			}
+
+			vs := make([]bool, slice.Len())
+
+			for i := range vs {
+				vs[i] = slice.Value(i)
+			}
+
+			if got, want := vs, tc.want; !reflect.DeepEqual(got, want) {
+				t.Fatalf("got=%v, want=%v", got, want)
+			}
+		})
+	}
+}
+
+func TestBooleanSliceOutOfBounds(t *testing.T) {
+	pool := memory.NewCheckedAllocator(memory.NewGoAllocator())
+	defer pool.AssertSize(t, 0)
+
+	values := []bool{true, false, true, false, true, false, true, false, true, false}
+
+	b := array.NewBooleanBuilder(pool)
+	defer b.Release()
+
+	for _, v := range values {
+		b.Append(v)
+	}
+
+	arr := b.NewArray().(*array.Boolean)
+	defer arr.Release()
+
+	slice := array.NewSlice(arr, 3, 8).(*array.Boolean)
+	defer slice.Release()
+
+	tests := []struct {
+		index int
+		panic bool
+	}{
+		{
+			index: -1,
+			panic: true,
+		},
+		{
+			index: 5,
+			panic: true,
+		},
+		{
+			index: 0,
+			panic: false,
+		},
+		{
+			index: 4,
+			panic: false,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run("", func(t *testing.T) {
+
+			if tc.panic {
+				defer func() {
+					e := recover()
+					if e == nil {
+						t.Fatalf("this should have panicked, but did not")
+					}
+					if got, want := e.(string), "arrow/array: index out of range"; got != want {
+						t.Fatalf("invalid error. got=%q, want=%q", got, want)
+					}
+				}()
+			} else {
+				defer func() {
+					if e := recover(); e != nil {
+						t.Fatalf("unexpected panic: %v", e)
+					}
+				}()
+			}
+
+			slice.Value(tc.index)
```

Comment:
perhaps store the value retuned by `Value(int)` and display it if `tc.panic` is `true`?
this way we make sure the compiler won't elide this piece of code?